python/ray/train/v2/_internal/execution/checkpoint/checkpoint_manager.py

```diff
@@ -81,63 +85,80 @@ def __init__(
     ):
         self._storage_context = storage_context
         self._checkpoint_config = checkpoint_config
+        self._num_reported_checkpoints = 0
+        self._num_pending_checkpoint_registrations = 0
+        self._condition = asyncio.Condition()
         super().__init__(checkpoint_config)
         # If the snapshot is found, the checkpoint manager will restore its state.
         self._maybe_load_state_from_storage()
 
     def register_checkpoint(self, checkpoint_result: _TrainingResult):
         """Register new checkpoint and add to bookkeeping.
 
-        This method will register a new checkpoint and add it to the internal
-        bookkeeping logic. This means the checkpoint manager will decide if
-        this checkpoint should be kept, and if older or worse performing
-        checkpoints should be deleted.
+        This method schedules an asyncio Task to register a new checkpoint
+        and add it to the internal bookkeeping logic. This means the
+        checkpoint manager will decide if this checkpoint should be kept,
+        and if older or worse performing checkpoints should be deleted.
 
         Args:
             checkpoint: Tracked checkpoint object to add to bookkeeping.
         """
-        self._latest_checkpoint_result = checkpoint_result
-
-        if self._checkpoint_config.checkpoint_score_attribute is not None:
-            # If we're ordering by a score, insert the checkpoint
-            # so that the list remains sorted.
-            _insert_into_sorted_list(
-                self._checkpoint_results,
-                checkpoint_result,
-                key=self._get_checkpoint_score,
-            )
-        else:
-            # If no metric is provided, just append (ordering by time of registration).
-            self._checkpoint_results.append(checkpoint_result)
-
-        results_to_delete = {}
-        if self._checkpoint_config.num_to_keep is not None:
-            # Delete the bottom (N - K) checkpoints
-            worst_results = set(
-                self._checkpoint_results[: -self._checkpoint_config.num_to_keep]
-            )
-            # Except for the latest checkpoint.
-            results_to_delete = worst_results - {self._latest_checkpoint_result}
-
-            # Update internal state before actually deleting them.
-            self._checkpoint_results = [
-                checkpoint_result
-                for checkpoint_result in self._checkpoint_results
-                if checkpoint_result not in results_to_delete
-            ]
+        asyncio.create_task(self.async_register_checkpoint(checkpoint_result))
+        self._num_pending_checkpoint_registrations += 1
```

Comment:
Why do we need to make this an async method? `after_report -> register_checkpoint` is already happening after all N workers have reported a checkpoint. There shouldn't be multiple callers to this `register_checkpoint` method.

Reply:
It's because `get_all_reported_checkpoints` waits for `self._num_reported_checkpoints == expected_num_checkpoints`, so the method that increments `self._num_reported_checkpoints` i.e. `register_checkpoint` needs to notify. I agree it's clunky and am open to solutions.

Reply:
Change so only the notification is in an asyncio task. This means:
* `build_result` doesn't need to wait for anything anymore. It will only be called after the checkpoints have been registered. 
* `get_all_reported_checkpoints` will still work - even if `controller.run` has exited, the asyncio event loop will still run the `notify_all` task which in turn will allow `get_all_reported_checkpoints` to return.